protocol/messenger_wallet.go

```diff
@@ -273,32 +249,72 @@ func (m *Messenger) syncKeypair(keypair *accounts.Keypair, fullKeypairSync bool,
 	rawMessage := common.RawMessage{
 		LocalChatID:         chat.ID,
 		ResendAutomatically: true,
+		MessageType:         protobuf.ApplicationMetadataMessage_SYNC_KEYPAIR,
 	}
 
-	if fullKeypairSync {
-		message, err := m.prepareSyncKeypairFullMessage(keypair)
-		if err != nil {
-			return err
-		}
+	message, err := m.prepareSyncKeypairMessage(keypair)
+	if err != nil {
+		return err
+	}
 
-		rawMessage.MessageType = protobuf.ApplicationMetadataMessage_SYNC_FULL_KEYPAIR
-		rawMessage.Payload, err = proto.Marshal(message)
-		if err != nil {
-			return err
+	rawMessage.Payload, err = proto.Marshal(message)
+	if err != nil {
+		return err
+	}
+
+	_, err = rawMessageHandler(ctx, rawMessage)
+	return err
+}
+
+// This function resolves which protobuf message needs to be sent depends on the passed `instance`
+//
+// If `instance` is of `*accounts.Keypair` type, we send `protobuf.SyncKeypair` message
+//
+// If `instance` is of `*accounts.SyncAccount` type:
+// - and if `KeyUID` is empty (means it's a watch only account) we send `protobuf.SyncAccount` message
+// - and if `KeyUID` is not empty (means the account belong to a keypair) we send `protobuf.SyncKeypair` message
+func (m *Messenger) resolveAndSyncKeypairOrJustWalletAccount(instance interface{}, rawMessageHandler RawMessageHandler) error {
+	if !m.hasPairedDevices() {
+		return nil
+	}
+
+	var (
+		clock    uint64
+		kpKeyUID string
+	)
+	_, chat := m.getLastClockWithRelatedChat()
+
+	switch obj := instance.(type) {
```

Comment:
I'm wondering whether we could just add some interface with `KeyUID` and `Clock` fields and avoid having `switch` here. Not a big deal, but it might have looked a bit simpler then.

Reply:
I agree that could be another option, but in that case we would need to update `syncWalletAccount` function as well. This way we use `obj` in the line `292`, cause in case of a watch only account we don't fetch it from the db in order to dispatch it, but use the `obj`.